cassandra/tablets.py

```diff
@@ -0,0 +1,99 @@
+# Experimental, this interface and use may change
+from threading import Lock
+
+class Tablet(object):
+    """
+    Represents a single ScyllaDB tablet.
+    It stores information about each replica, its host and shard, 
+    and the token interval in the format (first_token, last_token].
+    """
+    first_token = 0
+    last_token = 0
+    replicas = None
+
+    def __init__(self, first_token = 0, last_token = 0, replicas = None):
+        self.first_token = first_token
+        self.last_token = last_token
+        self.replicas = replicas
+
+    @staticmethod
+    def _is_valid_tablet(replicas):
+        return replicas is not None and len(replicas) != 0
+
+    @staticmethod
+    def from_row(first_token, last_token, replicas):
+        if Tablet._is_valid_tablet(replicas):
+            tablet = Tablet(first_token, last_token,replicas)
+            return tablet
+        return None
+
+# Experimental, this interface and use may change
+class Tablets(object):
+    _lock = None
+    _tablets = {}
+
+    def __init__(self, tablets):
+        self._tablets = tablets
+        self._lock = Lock()
+    
+    def get_tablet_for_key(self, keyspace, table, t):
+        tablet = self._tablets.get((keyspace, table), [])
+        if tablet == []:
+            return None
+        
+        id = bisect_left(tablet, t.value, key = lambda tablet: tablet.last_token)
+        if id < len(tablet) and t.value > tablet[id].first_token:
+            return tablet[id]
+        return None
+
+    def add_tablet(self, keyspace, table, tablet):
+        with self._lock:
+            tablets_for_table = self._tablets.setdefault((keyspace, table), [])
+
+            start = None
+            insert_idx = bisect_left(tablets_for_table, tablet.first_token, key = lambda t: t.first_token)
+            if insert_idx > 0 and insert_idx - 1 < len(tablets_for_table) and tablets_for_table[insert_idx - 1].last_token > tablet.first_token and tablets_for_table[insert_idx - 1].first_token < tablet.last_token:
+                insert_idx = insert_idx - 1
+                start = insert_idx
+            end = bisect_left(tablets_for_table, tablet.last_token, key = lambda t: t.last_token)
+            
+            if start != None:
```

Comment:
start is None when insert_idx == 0, in which case we skip the deletion, which is wrong if the new range overlaps with the first range.

Reply:
This is not the only case - in general it doesn't work if `tablets_for_table[insert_idx - 1]` doesn't overlap with `tablet` - it happens when `tablets_for_table[insert_idx].first_token == tablet.first_token` or of there is a gap in ranges so that `tablets_for_table[insert_idx - 1].last_token < tablet.first_token < tablets_for_table[insert_idx].first_token`